src/terrain/index.js

```diff
@@ -0,0 +1,157 @@
+const { getRandomTerrainFunction } = require('./terrainFunctions');
+const terrainGenerator = (boardBounds = { x: 100, y: 100 }, terrainBounds = { x: 5, y: 5 }, numberOfTerrainElements = 5) => {
+    let { mockArray, mockWidth, mockHeight } = createMockArray(boardBounds, terrainBounds);
+    const keyPoint = {
+        x: Math.floor(Math.random() * mockWidth),
+        y: Math.floor(Math.random() * mockHeight)
+    };
+    mockArray = generateTerrain(mockArray, keyPoint, numberOfTerrainElements, mockWidth, mockHeight);
+    return mockArray;
+};
+
+createMockArray = (boardBounds, terrainBounds) => {
+    const boardX = boardBounds.x;
+    const boardY = boardBounds.y;
+    const terrainX = terrainBounds.x;
+    const terrainY = terrainBounds.y;
+
+    const mockArray = [];
+    const mockWidth =  boardX/terrainX;
+    const mockHeight = boardY/terrainY;
+
+    const defaultInfo = {
+        filled: false,
+        north: true,
+        south: true,
+        east: true,
+        west: true
+    };
+
+    for(let i = 0; i < mockWidth; i++) {
+        mockArray[i] = [];
+        for (let j = 0; j < mockHeight; j++) {
+            const info = { ...defaultInfo };
+            if (i === 0) {
+                info.west = false;
+            } else if (i === mockWidth - 1) {
+                info.east = false;
+            }
+            if (j === 0) {
+                info.north = false;
+            } else if (j === mockHeight - 1) {
+                info.south = false;
+            }
+            mockArray[i].push(info);
+        }
+    }
+    return {
+        mockArray,
+        mockWidth,
+        mockHeight
+    };
+};
+
+const deepClone = (toCopy) => {
+    const toReturn = [];
+    toCopy.forEach((row, rowNum) => {
+        toReturn[rowNum] = [];
+        row.forEach(elem => {
+            toReturn[rowNum].push({ ...elem });
+        });
+    });
+    return toReturn;
+};
+
+
+const cleanUpBoard = (board) => {
+    for (let i = 0; i < board.length; i++) {
+        for (let j = 0; j < board[i].length; j++) {
+            board[i][j] = {
+                filled: board[i][j].filled,
+                north: board[i][j].north,
+                south: board[i][j].south,
+                east: board[i][j].east,
+                west: board[i][j].west
+            };
+        }
+    }
+};
+
+const verifyKeyPoint = (board, keyPoint, boardWidth, boardHeight) => {
```

Comment:
It's not immediately clear to me what the "key point" is. I guess overall for this stuff it seems like there is an algorithm to accomplish a goal of generating terrain, but it's hard to determine what that algorithm is from the code alone. It's kind of a mix of code and functions and conditions which ultimately make it harder to understand the actual process of what we're trying to do, which is randomly generate terrain. I don't really have a clear suggestion here or even a clear problem I can point to, I guess I'm just saying it's hard to find the actual important rules here and separate them from the code / functions

Reply:
yeah maybe we can discuss this further on the podcast, I'm not really sure how to clean this up in a way to make that part more clear